discovery_engine_core/core/src/engine.rs

```diff
@@ -289,16 +295,79 @@ where
     /// Process the feedback about the user reacting to a document.
     pub async fn user_reacted(&mut self, reacted: &UserReacted) -> Result<(), Error> {
         let mut stacks = self.stacks.write().await;
-        stacks
-            .get_mut(&reacted.stack_id)
-            .ok_or(Error::InvalidStackId(reacted.stack_id))?
-            .update_relevance(reacted.reaction);
+
+        // update relevance of stack if the reacted document belongs to one
+        if !reacted.stack_id.is_nil() {
+            stacks
+                .get_mut(&reacted.stack_id)
+                .ok_or(Error::InvalidStackId(reacted.stack_id))?
+                .update_relevance(reacted.reaction);
+        };
 
         self.ranker.log_user_reaction(reacted)?;
 
         rank_stacks(stacks.values_mut(), &mut self.ranker)
     }
 
+    /// Perform an active search with the given query parameters.
+    pub async fn active_search(
+        &mut self,
+        query: &str,
+        page: usize,
+        page_size: usize,
+    ) -> Result<Vec<Document>, Error> {
+        if query.trim().is_empty() {
+            return Err(Error::InvalidQuery(query.to_string()));
+        }
+        let EndpointConfig {
+            api_key,
+            api_base_url,
+            markets,
+            ..
+        } = &self.config;
+
+        let filter = &Filter::default().add_keyword(query);
+        let client = Client::new(api_key.clone(), api_base_url.clone());
+        let mut articles = Vec::new();
+        let mut errors = Vec::new();
+        for market in markets.read().await.iter() {
+            let news_query = NewsQuery {
+                market,
+                filter,
+                page_size,
+                page: Some(page),
+            };
+            match client.news(&news_query).await {
+                Ok(batch) => articles.extend(batch),
+                Err(err) => errors.push(Error::Client(err.into())),
+            };
+        }
+
+        let mut documents = Vec::new();
+        let stack_id = uuid::Uuid::nil().into(); // documents here not associated with a stack
+        for article in articles {
+            match self
+                .ranker
+                .compute_smbert(&article.title)
+                .map_err(Error::Ranker)
+                .and_then(|embedding| {
+                    document_from_article(article, stack_id, embedding).map_err(Error::Document)
+                }) {
+                Ok(doc) => documents.push(doc),
+                Err(err) => errors.push(err),
+            }
+        }
```

Comment:
This can be made more functional.
```rust
 let mut documents = articles                                                                     
            .into_iter()                                                                                 
            .filter_map(|article| {                                                                      
                self.ranker                                                                              
                    .compute_smbert(&article.title)                                                      
                    .map_err(Error::Ranker)                                                              
                    .and_then(|embedding| {                                                              
                        document_from_article(article, stack_id, embedding).map_err(Error::Document)     
                    })                                                                                   
                    .map_err(|e| errors.push(e))                                                         
                    .ok()                                                                                
            })                                                                                           
            .collect::<Vec<_>>();   
```

Reply:
nice :+1: